Source/Core/Common/BitField.h

```diff
@@ -133,21 +140,69 @@ struct BitField
   // so that we can use this within unions
   constexpr BitField() = default;
 
-  // We explicitly delete the copy assignment operator here, because the
-  // default copy assignment would copy the full storage value, rather than
-  // just the bits relevant to this particular bit field.
-  // Ideally, we would just implement the copy assignment to copy only the
-  // relevant bits, but we're prevented from doing that because the savestate
-  // code expects that this class is trivially copyable.
-  BitField& operator=(const BitField&) = delete;
-
-  DOLPHIN_FORCE_INLINE BitField& operator=(T val)
+  // The default copy-assignment copies the full storage value, rather than just the bits relevant
+  // to this particular BitField.  Ideally, we would just implement the copy assignment to copy only
+  // the relevant bits, but we're prevented from doing that because many classes containing a union
+  // with BitFields must be trivially copyable.  Savestate code is a great example.  At the same
+  // time, explicitly deleting copy-assignment would be undesireable, because all classes containing
+  // a union with BitFields would implicitly delete copy-assignment as well.  This would restrict
+  // simple things like array initialization, std::swap, std::fill, std::memcpy, etc.  The only real
+  // solution is to allow copy-assignment to exist, albeit with a warning so that code-review is not
+  // impossible.
+  DOLPHIN_DEPRECATED_MSG("If you can see this warning, you might be using BitField the wrong way.")
+  constexpr BitField& operator=(const BitField&) = default;
```

Comment:
Have you made sure that the deprecated message actually shows up?  When I was working on a similar thing that eventually ended up as dolphin-emu/hwtests#47, the warning didn't show up on GCC when `default` was used, but maybe I was doing something wrong?

Reply:
I have caught many unexpected behavior copy-assignments using the error message since I made that change to progress further.

Reply:
One interesting thing I've seen is that copy-assignment between BitFields with different template args does not seem to be an issue.  Would probably be best to still use .Value() just to be consistent.

Reply:
Hmm, I re-tested it, and both it and the standardized [`[[deprecated("Message")]]`](https://en.cppreference.com/w/cpp/language/attributes/deprecated) do generate warnings properly on GCC and Clang (even the version of GCC used by devkitpro).  But no warning is generated on MSVC, and Clang also generates a warning for the implicit usage.  See https://godbolt.org/z/6EWEMfvGe and the [pr-osx-universal buildlog](https://dolphin.ci/#/builders/26/builds/2935/steps/4/logs/stdio) near line 1242.

CC @neobrain

Reply:
I did not realize [[deprecated]] was a language standard.  I think I will get rid of the header I created and just use that.

Reply:
Clang Format is absolutely wrecking the [[deprecated("message")]] one.
[https://stackoverflow.com/questions/45740466/clang-format-setting-to-control-c-attributes](url)

Reply:
What exactly is the idea here? Is there going to be a bunch of "deprecated: ..." spam by default that can be ignored for "blessed" uses of this operator (i.e. those needed to copy unions), and someone needs to keep a close eye on making sure no new warnings slip through code review in the future?

Reply:
What exactly is the idea here? Is there going to be a bunch of "deprecated: ..." spam by default that can be ignored for "blessed" uses of this operator (i.e. those needed to copy unions), and someone needs to keep a close eye on making sure no new warnings slip through code review in the future?

Or is the hope to find a variant of this that only triggers the warning for explicit BitField-to-BitField copies but not with implicit copying through unions?

Reply:
FWIW, I'm not a huge fan of the deprecated tag being used in this way.  There's too much book keeping.  Feels like a design smell.

This is the only way I could see to do what you want, but it's very ugly in a _different_ way:

```
#include <algorithm>

template <typename Parent>
class Foo
{
public:
    Foo() = default;
    void Copy(const Foo<Parent>& other)
    {
        // Do relevant copying...
    }
    int i;
private:
    friend Parent;

    // rule of 5
    Foo& operator=(const Foo& other) = default;
    Foo& operator=(Foo&& other) = default;
    Foo(const Foo& other) = default;
    Foo(Foo&& other) = default;
};

union Bar
{
    Foo<Bar> m_foo;
    Foo<Bar> m_foo2;
};

int main()
{
    Bar bar1;
    Bar bar2;
    bar2.m_foo = bar1.m_foo;  // compiler error
    bar1 = bar2;  // works
    bar1.m_foo.Copy(bar2.m_foo);  // works
    std::swap(bar1, bar2);  // works
    return 0;
}
````